### PR TITLE
IGNITE-21157 Introduce transaction priorities

### PR DESCRIPTION
**Pull request by Flaugh24**
https://issues.apache.org/jira/browse/IGNITE-21157

Update WaitDieDeadlockPreventionPolicy - allow young transactions to wait if the current lock owner has a lower priority

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/transactions/src/test/java/org/apache/ignite/internal/tx/impl/ReadOnlyTransactionImplTest.java [hunk 1]
```diff
@@ -39,7 +40,7 @@ class ReadOnlyTransactionImplTest extends BaseIgniteAbstractTest {
     @Test
     void effectiveSchemaTimestampIsReadTimestamp() {
         HybridTimestamp readTimestamp = new HybridClockImpl().now();
-        UUID txId = TestTransactionIds.TRANSACTION_ID_GENERATOR.transactionIdFor(readTimestamp);
+        UUID txId = TestTransactionIds.TRANSACTION_ID_GENERATOR.transactionIdFor(readTimestamp, TxPriority.NORMAL);
```
Comment (ascherbakoff) [inline modules/transactions/src/test/java/org/apache/ignite/internal/tx/impl/ReadOnlyTransactionImplTest.java:43]: NORMAL should be the default priority then calling `transactionIdFor(ts).`

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/TxPriority.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        return byteValue;
+    }
+
+    /**
+     * Returns the priority from the byte value.
+     *
+     * @param byteValue the byte value.
+     * @return the priority.
+     */
+    static TxPriority fromPriority(byte byteValue) {
```
Comment (ascherbakoff) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/TxPriority.java:44]: static TxPriority fromPriority(byte byteValue)
->
static TxPriority fromUUID(UUID txid)

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/TxPriority.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+import org.apache.ignite.internal.tx.impl.WaitDieDeadlockPreventionPolicy;
+
+/**
+ * Transaction priority. The priority is used in {@link WaitDieDeadlockPreventionPolicy} to determine if a waiter should wait or abort.
+ */
+public enum TxPriority {
+    LOW((byte) 0),
+    NORMAL((byte) 1);
+    private final byte byteValue;
```
Comment (ascherbakoff) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/TxPriority.java:28]: byte value doesn't make sense here

File: modules/client/src/test/java/org/apache/ignite/client/fakes/FakeTxManager.java [hunk 4]
```diff
@@ -64,8 +65,14 @@ public InternalTransaction begin(HybridTimestampTracker tracker) {
         return begin(tracker, false);
     }
 
+
```
Comment (PakhomovAlexander) [inline modules/client/src/test/java/org/apache/ignite/client/fakes/FakeTxManager.java:68]:
```suggestion
```

File: modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java [hunk 5]
```diff
@@ -46,4 +70,28 @@ public static UUID transactionId(HybridTimestamp beginTimestamp, int nodeId) {
...
+
+    private static long combine(int nodeId, boolean priority) {
+        int priorityAsInt = priority ? 1 : 0;
+
+        // Shift the int 32 bits and combine with the boolean
+        return ((long) nodeId << 32) | priorityAsInt;
+    }
+
+    private static boolean encodePriority(TxPriority priority) {
+        return priority == TxPriority.NORMAL;
```
Comment (ascherbakoff) [inline modules/transactions/src/main/java/org/apache/ignite/internal/tx/TransactionIds.java:91]: Just use TxPriority.ordinal()

**Review by ascherbakoff — APPROVED** [review]

**Pull request merged**